File: packages/app/lib/features/groups/models/vector_clock.dart
import 'package:equatable/equatable.dart';

/// A vector clock for tracking causal ordering of messages in a group.
///
/// Maps each peer's device ID to the latest sequence number received from
/// that peer. Used for:
/// - Detecting missing messages during sync
/// - Establishing causal ordering (happens-before)
/// - Identifying which messages a peer is missing
class VectorClock extends Equatable {
  /// Map of {device_id: sequence_number}.
  ///
  /// Each entry represents the highest sequence number seen from that device.
  final Map<String, int> _clock;

  const VectorClock([Map<String, int>? clock]) : _clock = clock ?? const {};

  /// Create a vector clock from a map.
  factory VectorClock.fromMap(Map<String, int> map) {
    return VectorClock(Map<String, int>.from(map));
  }

  /// Get the sequence number for a specific device.
  ///
  /// Returns 0 if no messages have been received from the device.
  int operator [](String deviceId) => _clock[deviceId] ?? 0;

  /// Get all device IDs tracked by this clock.
  Set<String> get deviceIds => _clock.keys.toSet();

  /// Get the underlying map (immutable view).
  Map<String, int> toMap() => Map<String, int>.unmodifiable(_clock);

  /// Returns true if this clock has no entries.
  bool get isEmpty => _clock.isEmpty;

  /// Returns the number of entries in the clock.
  int get length => _clock.length;

  /// Increment the sequence number for a device.
  ///
  /// Returns a new [VectorClock] with the updated value.
  VectorClock increment(String deviceId) {
    final updated = Map<String, int>.from(_clock);
    updated[deviceId] = (updated[deviceId] ?? 0) + 1;
    return VectorClock(updated);
  }

  /// Set the sequence number for a device to a specific value.
  ///
  /// Returns a new [VectorClock] with the updated value.
  VectorClock set(String deviceId, int sequenceNumber) {
    final updated = Map<String, int>.from(_clock);
    updated[deviceId] = sequenceNumber;
    return VectorClock(updated);
  }

  /// Merge this clock with another, taking the maximum sequence number
  /// for each device.
  ///
  /// This is the standard vector clock merge operation used when
  /// receiving a message or syncing with a peer.
  VectorClock merge(VectorClock other) {
    final merged = Map<String, int>.from(_clock);
    for (final entry in other._clock.entries) {
      final current = merged[entry.key] ?? 0;
      if (entry.value > current) {
        merged[entry.key] = entry.value;
      }
    }
    return VectorClock(merged);
  }

  /// Returns true if this clock is causally before or equal to [other].
  ///
  /// A <= B iff for all device_id d: A[d] <= B[d]
  bool isBeforeOrEqual(VectorClock other) {
    for (final entry in _clock.entries) {
      if (entry.value > (other._clock[entry.key] ?? 0)) {
        return false;
      }
    }
    return true;
  }

  /// Returns true if this clock is strictly before [other].
  ///
  /// A < B iff A <= B and A != B
  bool isBefore(VectorClock other) {
    return isBeforeOrEqual(other) && this != other;
  }

  /// Returns true if this clock and [other] are concurrent
  /// (neither is causally before the other).
  bool isConcurrentWith(VectorClock other) {
    return !isBeforeOrEqual(other) && !other.isBeforeOrEqual(this);
  }

  /// Compute the set of messages that [other] is missing compared to this clock.
  ///
  /// Returns a map of {device_id: [list of missing sequence numbers]}.
  /// For each device where this clock is ahead of [other], the missing
  /// sequence numbers are listed.
  Map<String, List<int>> missingFrom(VectorClock other) {
    final missing = <String, List<int>>{};
    for (final entry in _clock.entries) {
      final otherSeq = other[entry.key];
      if (entry.value > otherSeq) {
        // The other peer is missing sequences (otherSeq+1) through entry.value
        missing[entry.key] = List.generate(
          entry.value - otherSeq,
          (i) => otherSeq + i + 1,
        );
      }
    }
    return missing;
  }

  /// Serialize to JSON-compatible map.
  Map<String, dynamic> toJson() =>
      _clock.map((key, value) => MapEntry(key, value));

  /// Deserialize from JSON map.
  factory VectorClock.fromJson(Map<String, dynamic> json) {
    return VectorClock(
      json.map((key, value) => MapEntry(key, (value as num).toInt())),
    );
  }

  @override
  List<Object?> get props => [_clock];

  @override
  String toString() => 'VectorClock($_clock)';
}

File: packages/headless-client/zajel/vector_clock.py
"""Vector clock for tracking causal ordering of messages in a group.

Maps each peer's device ID to the latest sequence number received from
that peer. Used for:
- Detecting missing messages during sync
- Establishing causal ordering (happens-before)
- Identifying which messages a peer is missing

This matches the Dart app's VectorClock (packages/app/lib/features/groups/models/vector_clock.dart).
"""

from __future__ import annotations

from typing import Any


class VectorClock:
    """A vector clock mapping device_id -> sequence_number.

    Each entry represents the highest sequence number seen from that device.
    """

    __slots__ = ("_clock",)

    def __init__(self, clock: dict[str, int] | None = None) -> None:
        self._clock: dict[str, int] = dict(clock) if clock else {}

    # ── Factory ────────────────────────────────────────────────

    @classmethod
    def from_map(cls, data: dict[str, Any]) -> VectorClock:
        """Deserialize from a dict (JSON-compatible).

        Values are cast to int to tolerate JSON numeric types.
        """
        return cls({k: int(v) for k, v in data.items()})

    # ── Accessors ──────────────────────────────────────────────

    def get(self, device_id: str) -> int:
        """Get the sequence number for a device (default 0)."""
        return self._clock.get(device_id, 0)

    @property
    def device_ids(self) -> set[str]:
        """All device IDs tracked by this clock."""
        return set(self._clock.keys())

    @property
    def is_empty(self) -> bool:
        """True if the clock has no entries."""
        return len(self._clock) == 0

    def to_map(self) -> dict[str, int]:
        """Serialize to a plain dict (JSON-compatible)."""
        return dict(self._clock)

    # ── Mutators ───────────────────────────────────────────────

    def increment(self, device_id: str) -> None:
        """Increment the counter for *device_id* by 1."""
        self._clock[device_id] = self._clock.get(device_id, 0) + 1

    def set(self, device_id: str, value: int) -> None:  # noqa: A003 — matches Dart API
        """Set the counter for *device_id* to *value*."""
        self._clock[device_id] = value

    def merge(self, other: VectorClock) -> None:
        """Element-wise max merge with *other* (in-place).

        After merging, for every device_id *d*:
            self[d] == max(self[d], other[d])
        """
        for device_id, seq in other._clock.items():
            current = self._clock.get(device_id, 0)
            if seq > current:
                self._clock[device_id] = seq

    # ── Causal ordering ────────────────────────────────────────

    def happened_before(self, other: VectorClock) -> bool:
        """Return True if this clock is *strictly* before *other*.

        self < other  iff  (for all d: self[d] <= other[d]) AND self != other
        """
        if self._clock == other._clock:
            return False
        for device_id, seq in self._clock.items():
            if seq > other._clock.get(device_id, 0):
                return False
        return True

    def is_concurrent(self, other: VectorClock) -> bool:
        """Return True if neither clock happened before the other.

        Concurrent means there exist device IDs where self is ahead
        AND device IDs where other is ahead (or equivalently,
        neither self <= other nor other <= self).
        """
        self_before_or_eq = self._is_before_or_equal(other)
        other_before_or_eq = other._is_before_or_equal(self)
        return not self_before_or_eq and not other_before_or_eq

    def _is_before_or_equal(self, other: VectorClock) -> bool:
        """Return True if self <= other (all entries)."""
        for device_id, seq in self._clock.items():
            if seq > other._clock.get(device_id, 0):
                return False
        return True

    # ── Sync ───────────────────────────────────────────────────

    def compute_missing(self, remote: VectorClock) -> dict[str, list[int]]:
        """Find messages that *remote* is missing compared to *self*.

        For each device_id where self is ahead of remote, returns the
        list of missing sequence numbers.

        Returns:
            {device_id: [missing_seq_1, missing_seq_2, ...]}
        """
        missing: dict[str, list[int]] = {}
        for device_id, local_seq in self._clock.items():
            remote_seq = remote.get(device_id)
            if local_seq > remote_seq:
                missing[device_id] = list(range(remote_seq + 1, local_seq + 1))
        return missing

    # ── Dunder methods ─────────────────────────────────────────

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, VectorClock):
            return NotImplemented
        return self._clock == other._clock

    def __repr__(self) -> str:
        return f"VectorClock({self._clock})"
